File: apps/api/app/schemas/sales.py
"""Pydantic request/response DTOs for Module 9's Sales, Payments, Returns & Refunds."""
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.db.enums import (
    InvoiceStatus,
    OrderPaymentStatus,
    PaymentMethod,
    QuotationStatus,
    RefundStatus,
    ReturnItemCondition,
    ReturnStatus,
    SaleStatus,
    SalesOrderStatus,
)


class LineItemRequest(BaseModel):
    plant_id: uuid.UUID | None = None
    inventory_id: uuid.UUID | None = None
    description: str | None = Field(None, max_length=500)
    quantity: int = Field(..., gt=0)
    unit_price: Decimal = Field(..., ge=0)
    discount_amount: Decimal = Field(Decimal("0"), ge=0)

    @model_validator(mode="after")
    def _at_most_one_reference(self) -> "LineItemRequest":
        if self.plant_id and self.inventory_id:
            raise ValueError("A line item may reference at most one of plant_id/inventory_id.")
        return self


# ------------------------------------------------------------------
# Quotations
# ------------------------------------------------------------------


class CreateQuotationRequest(BaseModel):
    branch_id: uuid.UUID
    customer_id: uuid.UUID
    items: list[LineItemRequest] = Field(..., min_length=1)
    tax_rate: float = Field(0.0, ge=0, le=1)
    header_discount: Decimal = Field(Decimal("0"), ge=0)
    valid_until: datetime | None = None
    note: str | None = Field(None, max_length=2000)


class QuotationStatusChangeRequest(BaseModel):
    status: QuotationStatus


class QuotationItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    quotation_id: uuid.UUID
    plant_id: uuid.UUID | None
    inventory_id: uuid.UUID | None
    description: str | None
    quantity: int
    unit_price: Decimal
    discount_amount: Decimal
    line_total: Decimal


class QuotationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    nursery_id: uuid.UUID
    branch_id: uuid.UUID
    customer_id: uuid.UUID
    status: QuotationStatus
    subtotal_amount: Decimal
    discount_amount: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    valid_until: datetime | None
    note: str | None
    created_by_user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


# ------------------------------------------------------------------
# Sales Orders
# ------------------------------------------------------------------


class CreateSalesOrderRequest(BaseModel):
    branch_id: uuid.UUID
    customer_id: uuid.UUID
    items: list[LineItemRequest] = Field(..., min_length=1)
    tax_rate: float = Field(0.0, ge=0, le=1)
    header_discount: Decimal = Field(Decimal("0"), ge=0)
    idempotency_key: str | None = Field(None, max_length=128)


class CancelOrderRequest(BaseModel):
    reason: str | None = Field(None, max_length=1000)


class OrderItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    sales_order_id: uuid.UUID
    plant_id: uuid.UUID | None
    inventory_id: uuid.UUID | None
    quantity: int
    unit_price: Decimal
    discount_amount: Decimal
    tax_amount: Decimal
    line_total: Decimal
    reservation_id: uuid.UUID | None


class SalesOrderResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    nursery_id: uuid.UUID
    branch_id: uuid.UUID
    customer_id: uuid.UUID
    quotation_id: uuid.UUID | None
    order_status: SalesOrderStatus
    payment_status: OrderPaymentStatus
    subtotal_amount: Decimal
    discount_amount: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    sale_id: uuid.UUID | None
    invoice_id: uuid.UUID | None
    created_by_user_id: uuid.UUID
    confirmed_at: datetime | None
    fulfilled_at: datetime | None
    cancelled_at: datetime | None
    cancel_reason: str | None
    created_at: datetime
    updated_at: datetime


# ------------------------------------------------------------------
# Sales (completed transactions) & Invoices
# ------------------------------------------------------------------


class SaleItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    sale_id: uuid.UUID
    plant_id: uuid.UUID | None
    inventory_id: uuid.UUID | None
    quantity: int
    unit_price: Decimal
    line_total: Decimal


class SaleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    nursery_id: uuid.UUID
    branch_id: uuid.UUID
    customer_id: uuid.UUID | None
    status: SaleStatus
    subtotal_amount: Decimal
    discount_amount: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    payment_method: str | None
    sold_by_user_id: uuid.UUID
    created_at: datetime


class InvoiceItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    invoice_id: uuid.UUID
    description: str
    quantity: int
    unit_price: Decimal
    line_total: Decimal


class InvoiceResponse(BaseModel):
    """`payment_status` is derived (sum of Payments vs. total_amount), not a stored column -- see `invoice_response()`."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    nursery_id: uuid.UUID
    branch_id: uuid.UUID
    customer_id: uuid.UUID
    invoice_number: str
    status: InvoiceStatus
    subtotal_amount: Decimal
    discount_amount: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    amount_paid: Decimal
    payment_status: OrderPaymentStatus
    due_date: datetime | None
    paid_at: datetime | None
    created_at: datetime


def invoice_response(invoice, *, amount_paid: Decimal) -> InvoiceResponse:
    if amount_paid >= Decimal(str(invoice.total_amount)) and amount_paid > 0:
        payment_status = OrderPaymentStatus.PAID
    elif amount_paid > 0:
        payment_status = OrderPaymentStatus.PARTIALLY_PAID
    else:
        payment_status = OrderPaymentStatus.UNPAID
    return InvoiceResponse(
        id=invoice.id, nursery_id=invoice.nursery_id, branch_id=invoice.branch_id, customer_id=invoice.customer_id,
        invoice_number=invoice.invoice_number, status=invoice.status, subtotal_amount=invoice.subtotal_amount,
        discount_amount=invoice.discount_amount, tax_amount=invoice.tax_amount, total_amount=invoice.total_amount,
        amount_paid=amount_paid, payment_status=payment_status, due_date=invoice.due_date, paid_at=invoice.paid_at,
        created_at=invoice.created_at,
    )


# ------------------------------------------------------------------
# Payments
# ------------------------------------------------------------------


class RecordPaymentRequest(BaseModel):
    amount: Decimal = Field(..., gt=0)
    method: PaymentMethod
    reference: str | None = Field(None, max_length=100)


class PaymentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    invoice_id: uuid.UUID
    amount: Decimal
    method: str
    reference: str | None
    received_by_user_id: uuid.UUID
    received_at: datetime


# ------------------------------------------------------------------
# Returns & Refunds
# ------------------------------------------------------------------


class ReturnItemRequest(BaseModel):
    sale_item_id: uuid.UUID
    quantity: int = Field(..., gt=0)
    restock: bool = True
    condition: ReturnItemCondition = ReturnItemCondition.RESALABLE


class CreateReturnRequest(BaseModel):
    customer_id: uuid.UUID
    reason: str | None = Field(None, max_length=2000)
    items: list[ReturnItemRequest] = Field(..., min_length=1)


class RejectReturnRequest(BaseModel):
    reason: str | None = Field(None, max_length=1000)


class ReturnItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    return_id: uuid.UUID
    sale_item_id: uuid.UUID
    quantity: int
    restock: bool
    condition: ReturnItemCondition
    line_refund_amount: Decimal


class ReturnResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    nursery_id: uuid.UUID
    branch_id: uuid.UUID
    sale_id: uuid.UUID
    customer_id: uuid.UUID
    status: ReturnStatus
    reason: str | None
    requested_by_user_id: uuid.UUID
    processed_by_user_id: uuid.UUID | None
    processed_at: datetime | None
    created_at: datetime


class ProcessRefundRequest(BaseModel):
    branch_id: uuid.UUID
    amount: Decimal = Field(..., gt=0)
    method: PaymentMethod
    return_id: uuid.UUID | None = None
    invoice_id: uuid.UUID | None = None
    sale_id: uuid.UUID | None = None
    reference: str | None = Field(None, max_length=100)


class RefundResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    nursery_id: uuid.UUID
    branch_id: uuid.UUID
    return_id: uuid.UUID | None
    invoice_id: uuid.UUID | None
    sale_id: uuid.UUID | None
    amount: Decimal
    method: PaymentMethod
    status: RefundStatus
    reference: str | None
    processed_by_user_id: uuid.UUID
    processed_at: datetime | None
    created_at: datetime


# ------------------------------------------------------------------
# Reporting
# ------------------------------------------------------------------


class SalesReportResponse(BaseModel):
    sale_count: int
    total_revenue: float
    total_tax: float
    total_discount: float
    average_sale_value: float


class RevenueReportRow(BaseModel):
    date: str
    revenue: float

File: apps/web/test/fixtures/sales.ts
import type {
  InvoiceItemResponse,
  InvoiceResponse,
  OrderItemResponse,
  PageQuotationResponse,
  PageRefundResponse,
  PageReturnResponse,
  PageSalesOrderResponse,
  PageSaleResponse,
  PaymentResponse,
  QuotationItemResponse,
  QuotationResponse,
  RefundResponse,
  ReturnItemResponse,
  ReturnResponse,
  RevenueReportRow,
  SaleItemResponse,
  SaleResponse,
  SalesOrderResponse,
  SalesReportResponse,
} from "@/lib/api/sales";

/** Shared fixtures for 7J Sales/Quotation/Order/Invoice/Return/Refund tests -- mirrors test/fixtures/inventory.ts's pattern. */

const BRANCH_ID = "44444444-4444-4444-4444-444444444444";
const CUSTOMER_ID = "cust-11111111-1111-1111-1111-111111111101";

export function makeQuotation(overrides: Partial<QuotationResponse> = {}): QuotationResponse {
  return {
    id: "quote-01",
    nursery_id: "22222222-2222-2222-2222-222222222222",
    branch_id: BRANCH_ID,
    customer_id: CUSTOMER_ID,
    status: "draft",
    subtotal_amount: "100.00",
    discount_amount: "0.00",
    tax_amount: "8.00",
    total_amount: "108.00",
    valid_until: null,
    note: null,
    created_by_user_id: "11111111-1111-1111-1111-111111111111",
    created_at: "2026-08-01T00:00:00Z",
    updated_at: "2026-08-01T00:00:00Z",
    ...overrides,
  };
}

export function makeQuotationPage(items: QuotationResponse[] = [makeQuotation()]): PageQuotationResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeQuotationItem(overrides: Partial<QuotationItemResponse> = {}): QuotationItemResponse {
  return {
    id: "quote-item-01",
    quotation_id: makeQuotation().id,
    plant_id: null,
    inventory_id: "inv-01",
    description: "4in nursery pots",
    quantity: 10,
    unit_price: "10.00",
    discount_amount: "0.00",
    line_total: "100.00",
    ...overrides,
  };
}

export function makeSalesOrder(overrides: Partial<SalesOrderResponse> = {}): SalesOrderResponse {
  return {
    id: "order-01",
    nursery_id: "22222222-2222-2222-2222-222222222222",
    branch_id: BRANCH_ID,
    customer_id: CUSTOMER_ID,
    quotation_id: null,
    order_status: "draft",
    payment_status: "unpaid",
    subtotal_amount: "100.00",
    discount_amount: "0.00",
    tax_amount: "8.00",
    total_amount: "108.00",
    sale_id: null,
    invoice_id: null,
    created_by_user_id: "11111111-1111-1111-1111-111111111111",
    confirmed_at: null,
    fulfilled_at: null,
    cancelled_at: null,
    cancel_reason: null,
    created_at: "2026-08-01T00:00:00Z",
    updated_at: "2026-08-01T00:00:00Z",
    ...overrides,
  };
}

export function makeSalesOrderPage(items: SalesOrderResponse[] = [makeSalesOrder()]): PageSalesOrderResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeOrderItem(overrides: Partial<OrderItemResponse> = {}): OrderItemResponse {
  return {
    id: "order-item-01",
    sales_order_id: makeSalesOrder().id,
    plant_id: null,
    inventory_id: "inv-01",
    quantity: 10,
    unit_price: "10.00",
    discount_amount: "0.00",
    tax_amount: "8.00",
    line_total: "108.00",
    reservation_id: null,
    ...overrides,
  };
}

export function makeSale(overrides: Partial<SaleResponse> = {}): SaleResponse {
  return {
    id: "sale-01",
    nursery_id: "22222222-2222-2222-2222-222222222222",
    branch_id: BRANCH_ID,
    customer_id: CUSTOMER_ID,
    status: "completed",
    subtotal_amount: "100.00",
    discount_amount: "0.00",
    tax_amount: "8.00",
    total_amount: "108.00",
    payment_method: "cash",
    sold_by_user_id: "11111111-1111-1111-1111-111111111111",
    created_at: "2026-08-01T00:00:00Z",
    ...overrides,
  };
}

export function makeSalePage(items: SaleResponse[] = [makeSale()]): PageSaleResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeSaleItem(overrides: Partial<SaleItemResponse> = {}): SaleItemResponse {
  return {
    id: "sale-item-01",
    sale_id: makeSale().id,
    plant_id: null,
    inventory_id: "inv-01",
    quantity: 10,
    unit_price: "10.00",
    line_total: "100.00",
    ...overrides,
  };
}

export function makeInvoice(overrides: Partial<InvoiceResponse> = {}): InvoiceResponse {
  return {
    id: "invoice-01",
    nursery_id: "22222222-2222-2222-2222-222222222222",
    branch_id: BRANCH_ID,
    customer_id: CUSTOMER_ID,
    invoice_number: "INV-0001",
    status: "sent",
    subtotal_amount: "100.00",
    discount_amount: "0.00",
    tax_amount: "8.00",
    total_amount: "108.00",
    amount_paid: "0.00",
    payment_status: "unpaid",
    due_date: "2026-09-01",
    paid_at: null,
    created_at: "2026-08-01T00:00:00Z",
    ...overrides,
  };
}

export function makeInvoiceItem(overrides: Partial<InvoiceItemResponse> = {}): InvoiceItemResponse {
  return {
    id: "invoice-item-01",
    invoice_id: makeInvoice().id,
    description: "4in nursery pots",
    quantity: 10,
    unit_price: "10.00",
    line_total: "100.00",
    ...overrides,
  };
}

export function makePayment(overrides: Partial<PaymentResponse> = {}): PaymentResponse {
  return {
    id: "payment-01",
    invoice_id: makeInvoice().id,
    amount: "50.00",
    method: "cash",
    reference: null,
    received_by_user_id: "11111111-1111-1111-1111-111111111111",
    received_at: "2026-08-05T00:00:00Z",
    ...overrides,
  };
}

export function makeReturn(overrides: Partial<ReturnResponse> = {}): ReturnResponse {
  return {
    id: "return-01",
    nursery_id: "22222222-2222-2222-2222-222222222222",
    branch_id: BRANCH_ID,
    sale_id: makeSale().id,
    customer_id: CUSTOMER_ID,
    status: "requested",
    reason: "Wrong item received",
    requested_by_user_id: "11111111-1111-1111-1111-111111111111",
    processed_by_user_id: null,
    processed_at: null,
    created_at: "2026-08-06T00:00:00Z",
    ...overrides,
  };
}

export function makeReturnPage(items: ReturnResponse[] = [makeReturn()]): PageReturnResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeReturnItem(overrides: Partial<ReturnItemResponse> = {}): ReturnItemResponse {
  return {
    id: "return-item-01",
    return_id: makeReturn().id,
    sale_item_id: makeSaleItem().id,
    quantity: 2,
    restock: true,
    condition: "resalable",
    line_refund_amount: "20.00",
    ...overrides,
  };
}

export function makeRefund(overrides: Partial<RefundResponse> = {}): RefundResponse {
  return {
    id: "refund-01",
    nursery_id: "22222222-2222-2222-2222-222222222222",
    branch_id: BRANCH_ID,
    return_id: makeReturn().id,
    invoice_id: null,
    sale_id: null,
    amount: "20.00",
    method: "cash",
    status: "completed",
    reference: null,
    processed_by_user_id: "11111111-1111-1111-1111-111111111111",
    processed_at: "2026-08-07T00:00:00Z",
    created_at: "2026-08-07T00:00:00Z",
    ...overrides,
  };
}

export function makeRefundPage(items: RefundResponse[] = [makeRefund()]): PageRefundResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeSalesReport(overrides: Partial<SalesReportResponse> = {}): SalesReportResponse {
  return { sale_count: 12, total_revenue: 1240.5, total_tax: 96.0, total_discount: 20.0, average_sale_value: 103.38, ...overrides };
}

export function makeRevenueReportRow(overrides: Partial<RevenueReportRow> = {}): RevenueReportRow {
  return { date: "2026-08-01", revenue: 108.0, ...overrides };
}
